engineering-plans/0004-subgraph-grafting.md

```diff
@@ -0,0 +1,143 @@
+           probably be faster to copy the data with `create table new.table
+           as table old.table`; that requires that we manually create
+           constraints and indexes after copying the data. As that requires
+           a deeper change in the code that creates the database schema for
+           a subgraph, we'll leave that as a possible improvement
+        3. copy dynamic data sources from the source subgraph. The copying
+           will create new `DynamicEthereumContractDataSource` entities,
+           and all their child entities, by creating a new id for the
+           dynamic data source and adjusting the id's of child entities
+           accordingly
+        4. rewind the new subgraph to the desired block by reverting to
+           that block (both data and metadata)
+        5. set the `latestEthereumBlockHash` and
+           `latestEthereumBlockNumber` of the new subgraph's
+           `SubgraphDeployment` to the desired block
+    - For `reuse`:
+        1. change the `deployment` field of the source subgraph's dynamic
+           data sources to point at the new subgraph
+        2. rename the database schema `sgdNNN` for the source subgraph's
+           data to the target subgraph's name using `alter schema sgdOLD
+           rename to sgdNEW`
+        3. rewind the new subgraph to the desired block
+        4. remove the metadata for the old subgraph
+        5. delete the entry for `sgdOLD` from `deployment_schemas` Strictly
+           speaking, we also need to remove the entry for `sgdOLD` from
+           each `graph-node`'s `Store.subgraph_cache` and
+           `Store.storage_cache` but do not have a mechanism in place for
+           that. We should check, but leaving these entries in the cache
+           should be ok; it might lead to less than helpful error messages
+           when trying to query the old subgraph but should cause no harm
+           beyond that
+
+## Tests
+
+Details TBD, but the tests need to cover at least subgraph manifests
+without a `deployment` and each of the supported `modes`
+
+## Migration
+
+This is purely additive and requires no migration.
+
+## Documentation
+
+The `deployment` field in subgraph manifests will be documented [in the
+public
+docs](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest)
+and the [manifest reference](https://github.com/graphprotocol/graph-node/blob/master/docs/subgraph-manifest.md)
+
+## Implementation Plan
+
+The plan is described above under 'Implementation'
+
+## Open Questions
+
+- Should subgraphs deployed with `copy` or `reuse` be marked somehow as it
+  will be very hard to recreate them in other `graph-node` installations?
+  Recreating such a subgraph would require deploying the source subgraph,
+  letting it index past the source block and then deploying the target
+  subgraph. Since the source subgraph could itself be the result of
+  grafting, it might be necessary to perform these operations recursively
```

Comment:
Recreating a grafted subgraph on a different graph node is indeed tricky. I have no good ideas here yet...

Reply:
It does defeat the entire point if the node doesn't have the base subgraph synced to the required block. In those cases the deployment should just be rejected imo.

Reply:
I disagree that this is hard to recreate - subgraphs grafted upon should be on IPFS. That's what deploying means in The Network.

I also disagree that it defeats the entire point if the subgraph is not synced on a particular deployment. It may be synced on some nodes and not others, or synced locally and not in production. Quick syncing is not the only use for grafting, and quick syncing may only take effect in some environments but that doesn't mean it should fail in others where there is less benefit.

No, subgraphs deployed with `copy` or `reuse` should not be "marked somehow".

Reply:
What you are suggesting would require quite a bit more machinery: when a grafted subgraph is deployed, we would first need to (recursively) deploy its source, let that sync to the desired block (what if it fails before then?), and then trigger the deploy of the graft. We'd then also need to handle reverts across the grafting block - I should clarify this, but the behavior I plan on is that the revert would just happen within the data and logic of the grafted subgraph.

The main use case for this feature is to fix subgraph errors quickly, not as a general feature for code management. When it's planned, a subgraph writer can achieve the same effect as grafting with simple logic in their mappings, which is also much more flexible as the change can be triggered on more than just a block number. Grafting is really for situations where you have already synced for a long period of time, discover a bug in your mapping, and can address that by resyncing only a small portion of the subgraph rather than starting over.